Information flow (#346)
ietf-wg-privacypass/base-drafts: draft-ietf-privacypass-architecture.md

Attestation context:
: The interactions and set of information shared between
the Client and Attester only, for the purposes of attesting the vailidity of

> the Client for the purposes of completing the issuance protocol. This context

Reviewer: Two "for the purposes of..." here?

use any of these Attesters to issue and redeem tokens for the Origin.

The mechanisms for establishing trust between each entity in this arrangement
are deployment specific. For example, in settings where Clients interact with

> Issuers transitively through an Attester, Attesters and Issuers might use

Reviewer: ```suggestion
Issuers through an Attester, Attesters and Issuers might use
```

Any extension to the Privacy Pass protocol MUST adhere to the guidelines
specified in {{issuer-role}} for managing Issuer public key data.

## Information Flow {#flow}

The end-to-end run of redemption and issuance protocol involves information
flowing between Issuer, Origin, Client, and Attester. That information can
have implications on the privacy goals that Privacy Pass aims to provide
as outlined in {{privacy-and-trust}}. In this section, we describe the flow
of information between each party. How this information affects the privacy
goals in particular deployment models is further discussed in {{deployment}}.

### Token Challenge Flow {#challenge-flow}

To use Privacy Pass, Origins choose an Issuer from which they are willing to
accept tokens. Origins then construct a token challenge using this specified

> Issuer and the redemption context it shares with the Client. This token

Author: I think we need to keep redemption context here, since that's the information the Origin has available to it to construct a token challenge. I like the other suggestion to explain why issuance secrecy is important.

> Issuer and the redemption context it shares with the Client. This token

Reviewer 2: I still don't understand whether "redemption context" refers to data or something else though.

> Issuer and the redemption context it shares with the Client. This token

Author: It refers to _anything that the origin can see during redemption_ (defined earlier in the doc).

> Issuer and the redemption context it shares with the Client. This token

Author: (Unresolving) @martinthomson does the improved description in #343 work?

> Issuer and the redemption context it shares with the Client. This token

Reviewer 2: ```suggestion
Issuer and information from the redemption context it shares with the Client. This token
```

challenge is then delivered to a Client. The token challenge conveys
information about the Issuer and the redemption context, such as whether the
Origin desires a per-Origin or cross-Origin token.

> ### Attestation and Issuance Flow {#issue-flow}

Author: Yeah, I think so, since they're both happening in issuance.

Clients use the issuance protocol to produce a token bound to a token
challenge. In doing so, there are several ways in which the issuance protocol
contributes information to the attestation context and issuance context.
For example, a token request may contribute information to the attestation
or issuance contexts as described below.

- Issuance protocol. The type of issuance protocol can contribute information

> about the Issuer's capabilities to the attestation or issuance contexts.

Author: > Why would the issuance protocol choice leak information to the issuance context? That's not immediately obvious.

As an example, maybe an Issuer only supports the rate-limiting issuance protocol for one origin. This would be pretty silly, but it would be possible, and as a result, the issuance protocol immediately reveals unwanted information to the issuer.

> I think that what you are saying is that the choice made by the Origin regarding the issuance protocol might be used to carry information from the Origin to the Issuer if the Client requests issuance when it receives a challenge. If the issuance protocol choice is exposed to the Attester, then it too learns this information. If there are multiple options available, some information about the Client can be encoded in the choice between those options.

That's a good clarification, yeah. I'll include it.

- Issuer identity or configuration. Information about the Issuer, such as its
identity or the public key used to validate tokens it creates, can be revealed

> during issuance and contribute to the attestation or issuance contexts.

Author: I relaxed this to be more general and just say "issuer configuration," as that's generally what the token challenge conveys.

- Attestation information. The issuance protocol can contribute information to
the issuance context based on what attestation procedure the Issuer uses to
trust a token request. In particular, a token request that is validated by
a given Attester means that the Client which generated the token request must

> be capable of the completing the designated attestation procedure.

Reviewer 2: This is necessary information, isn't it?  We really do need the Issuer to be convinced that the Client passed whatever tests the Attester set.

The other aspect to consider is what unwanted information might be revealed here.  If the Attester would accept a Client that meets criterion A, B, or C, then it should not be the case that the Issuer learns which of A, B, or C was used.  That seems fine.  However, if there are multiple valid Attesters, I don't see a way to avoid the Issuer learning which Attester was used.  The choice of Attester therefore says something about the client.  There is a transitory flow of information here though: if Attester A supports method A and Attester B supports method B, then the choice of Attester indirectly reveals the choice of method.

> be capable of the completing the designated attestation procedure.

Author: This list isn't meant to be exhaustive, as I don't expect us to be able to list all exhaustive things. It contains the what seem to me to be the obvious things that might be revealed, and the type of attestation that's done is one of those obvious things. It is necessarily revealed to the issuer through the process here.

> be capable of the completing the designated attestation procedure.

Reviewer 2: Obvious or not, it is still worth spelling out explicitly.  These are points where information leaks as a direct result of the architectural choices.

- Origin information. The issuance protocol can contribute information about
the Origin that challenged the Client in {{challenge-flow}}. In particular,
a token request designated for a specific Issuer means that the resulting token

> is for an Origin that trusts the specified Issuer.

Reviewer 2: I think that you might infer that the Client had received a challenge from one of the Origins that trust that Issuer, but you can't guarantee that this is the case.  Clients could proactively retrieve tokens.  Those tokens wouldn't be bound to something the Origin cares about necessarily, so maybe a per-Origin token doesn't fit this cleanly, but it does seem possible.

Author: Good call out. I've clarified that it's only a possibility.

Moreover, a token response may contribute information to the attestation
or issuance contexts as described below.

- Origin information. The issuance protocol can contribute information about
the Origin in how it responds to a token request. For example, if an Issuer
learns the Origin during issuance and is also configured to respond in some way
on the basis of that information, that response can reveal information to the

> Attester.

Reviewer 2: I'll repeat my previous complaint.  The Client interacting with the Issuer *via the Attester* is something that is assumed far too often in the document.  This needs to be conditional at a minimum.  Also, in those protocols that encapsulate the Client-Issuer exchange, the information that leaks is only based on traffic analysis.

Author: This is fair criticism. It's noted that this is an option in the overview section, but I don't think it's obvious enough. I'll try to clarify.

as outlined in {{privacy-and-trust}}. In this section, we describe the flow
of information between each party. How this information affects the privacy
goals in particular deployment models is further discussed in {{deployment}}.

### Token Challenge Flow {#challenge-flow}

To use Privacy Pass, Origins choose an Issuer from which they are willing to
accept tokens. Origins then construct a token challenge using this specified
Issuer and the redemption context it shares with the Client. This token
challenge is then delivered to a Client. The token challenge conveys
information about the Issuer and the redemption context, such as whether the
Origin desires a per-Origin or cross-Origin token. Any entity that sees
the token challenge might learn things about the Client as known to the Origin.
This is why input secrecy is a requirement for issuance protocols, as it
ensures that the challenge is not directly available to the Issuer.

### Attestation and Issuance Flow {#issue-flow}

Clients use the issuance protocol to produce a token bound to a token
challenge. In doing so, there are several ways in which the issuance protocol

> contributes information to the attestation context and issuance context.

Reviewer 2: I'm still not convinced by this being a single section.  The attestation context only gains information from the issuance context transitively.  That is, if you need to use attestation method W to execute issuance protocol X and you need issuance protocol X to get token Y to answer challenge Z, then you can trace the way from Z to W, but each step is its own and the client is more or less in control of whether to proceed in each, provided that it understands what information carries over.

Author: The attestation context does sometimes gain information explicitly from issuance, as in the rate limited protocol. I'll try to clarify that in the definitions.

Any extension to the Privacy Pass protocol MUST adhere to the guidelines
specified in {{issuer-role}} for managing Issuer public key data.

## Information Flow {#flow}

> The end-to-end run of redemption and issuance protocol involves information

Reviewer: ```suggestion
The end-to-end process of redemption and issuance protocols involves information
```

This is why input secrecy is a requirement for issuance protocols, as it
ensures that the challenge is not directly available to the Issuer.

### Attestation Flow {#attestation-flow}

Clients interact with the Attester to prove that they meet some required
set of properties. In doing so, Clients contribute information to the
attestation context, which might include sensitive information such as
application-layer identities, IP addresses, and so on. Clients can choose
whether or not to contribute this information based on local policy or
preference.

### Issuance Flow {#issue-flow}

Clients use the issuance protocol to produce a token bound to a token
challenge. In doing so, there are several ways in which the issuance protocol
contributes information to the attestation or issuance contexts. For example, a
token request may contribute information to the attestation or issuance
contexts as described below.

- Issuance protocol. The type of issuance protocol can contribute information
about the Issuer's capabilities to the attestation or issuance contexts, as
well as the capabilities of a given Client. For example, if a Client is

> presented with multuiple issuance protocol options, then the choice of which

Reviewer 2: typo: mult~~u~~iple

> issuance protocol to use can contrbute information about the Client's

Reviewer 2: typo: contr[i]bute

a token request designated for a specific Issuer might imply that the resulting
token is for an Origin that trusts the specified Issuer. However, this is not
always true, as some token requests can correspond to cross-Origin tokens,
i.e., they are tokens that would be accepted at any Origin that accepts the
cross-Origin token.

Moreover, a token response may contribute information to the issuance
attestation or contexts as described below.

- Origin information. The issuance protocol can contribute information about
the Origin in how it responds to a token request. For example, if an Issuer
learns the Origin during issuance and is also configured to respond in some way
on the basis of that information, and the Client interacts with the Issuer
transitively through the Attester, that response can reveal information to the
Attester.
- Token. The token produced by the issuance protocol can contain information
from the issuance context. In particular, depending on the issuance protocol,
tokens can contain public or private metadata, and Issuers can choose that
metadata on the basis of information in the issuance context.

> In general, exceptional cases in the issuance protocol, such as when either the

Reviewer 2: I suggest that you avoid saying "In general" before "exceptional".
